### ISSUE TITLE
Return types of `c_ptrTo()` and `c_addrOf()` as applied to Chapel classes

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#22478]
As a follow-up to #22367, @twesterhout asked an interesting question which hadn't occurred to me before regarding the return types of `c_ptrTo()` and `c_addrOf()` as applied to class variables.  Specifically a declaration like:

```chapel
class C {}
var myC: unmanaged C = new unamanged C();
```

results in two values, an object `C` and a pointer to the object `C`.  Yet we refer to both of those values collectively as `unmanaged C`, so we might think of `c_ptrTo()` and `c_addrOf()` as both returning `c_ptr(unmanaged C)`.  Yet the values represented by these types are quite different—one is a C pointer to the object itself, the other is a C pointer to the Chapel pointer to the object.

What can we do to improve upon this?  Here are some ideas I'm not sure I'm so wild about:

* We could have `c_ptrTo(myC)` return `c_ptr(C)` where in that context `C` would be interpreted as the object itself rather than the pointer to the object, while `c_addrOf(myC)` might return `c_ptr(unmanaged C)`.  This has the downside of treating `C` somewhat differently than we traditionally have (which is to say, as a class type with generic management).
* We could have `c_ptrTo(myC)` return `c_ptr(unmanaged C)` and for `c_addrOf(myC)` return `c_ptr(c_ptr(unmanaged C))`—this feels a little funny to me in that I tend to think of `unmanaged C` as being the type of the pointer rather than the class itself, so it feels more like a triple indirection rather than a double indirection...

My best answer at this point might be that `c_addrOf()` potentially just returns a `c_void_ptr` for types that don't have an "obvious" C type, like the Chapel pointer to the class, an array, string, or bytes metadata record, etc.  Essentially, make it intentionally imprecise because we don't have a better Chapel type for describing the type in question.  Alternatively, we could make it a `c_ptr` to a Chapel-specific implementation type whose definition (and name?) may change over time.  That would suggest that `c_ptrTo(myC)` would still return `c_ptr(unmanaged C)` I suppose, which still seems a little weird (in the sense that it seems like a double indirection of a type to me).

This topic is loosely related to some of the comments on https://github.com/chapel-lang/chapel/issues/22341 which assert the importance of being able to convert Chapel class to a C pointer and back again (which is precisely what Tom is trying to do in this situation).

**Comment by jeremiah-corrado** [comment]
> Based on this I would actually prefer c_addrOf, which gets the thing on the stack, to be a c_ptr(unmanaged C), while c_ptrTo for the thing on the heap would be a c_void_ptr. 

To me, this feels consistent with the direction we went with strings and bytes. However, I'm not quite sure what the use case is for a `c_ptr(unmanaged C)`?

Based on the discussion in that other thread (#22341), it sounds like we need the ability to get a `c_void_ptr` to the class instance in order to convert back and forth between C and Chapel. I'd advocate for `c_ptrTo` taking that role regardless of what behavior we choose for `c_addrOf`.

**Comment by bradcray** [comment]
> Would a user doing this want to have a Chapel c_ptr to the stack thing, or the heap instance?

The heap instance I think (since the stack thing is arguably more ephemeral than the object itself).

**Comment by riftEmber** [comment]
Is there a way for a user to go from a `c_void_ptr` to the heap instance (provided by `c_ptrTo` say) back to a Chapel `unamanged C`?

**Comment by bradcray** [comment]
Yes, see https://matrix.to/#/!PBYDSerrfYujeStENM:gitter.im/$-IyECrexBOxZrDRrkWbxxYmd9PgjpqWlLksIzqA_z9I?via=gitter.im&via=matrix.org&via=gnome.org .  This also relates to some of the discussion on https://github.com/chapel-lang/chapel/issues/22341

Maybe turning this issue into more of a practical question:  How are the return types of these routines documented or described today for various types?

**Comment by twesterhout** [comment]
As a user I want to second @riftEmber's suggestion to have `c_addrOf` return a `c_void_ptr` when applied to an object of a class type. Having both `c_addrOf` and `c_ptrTo` return `c_ptr(unmanaged C)` contributed greatly to my confusion about their behavior... Had `c_addOf(x)` returned a `c_void_ptr` instead, it would've indicated to me that I shouldn't have expected to be able to reconstruct an `unmanaged C` from it when `x` went out of scope.

**Comment by riftEmber** [comment]
Thanks for the link, I see how my proposed types wouldn't help with that.

> How are the return types of these routines documented or described today for various types?

In the general case, either `c_ptrTo(x)` or `c_addrOf(x)` will return a `c_ptr(x)`. The only cases we do something smarter for are `c_ptrTo` on an array, which returns `c_ptr(eltType)`, and `c_ptrTo` on a string or bytes, which returns `c_ptr(c_uchar)`.

As `c_ptrTo` and `c_addrOf` do something very different from each other when given a class instance, it makes a lot of sense to me they should return different types for it.

@twesterhout / @bradcray :

I definitely see how it is misleading to be handed a `c_ptr(unmanaged C)` that can't be dereferenced into an `unmanaged C`. The sticking point for me here is my interpretation that an `unmanaged/owned/shared C` is the thing on the stack, and the thing on the heap is "something else" internal to Chapel. By that definition it makes sense that the `unmanaged C` is gone by the time you try to dereference it -- really what you want to dereference is the thing on the heap. So to take that to its conclusion, it would make more sense to me if you did `c_ptrTo(x)`, got back a `c_void_ptr` (since I am asserting the heap thing is not an "`unmanaged C`"), and then later down the line had to somehow create a new `unmanaged C` pointing to that `c_void_ptr` if you wanted to use it as a Chapel type again.

All that said, my interpretation of types of the stack record/heap instance is not necessarily the only "correct" one; one could also say the `unmanaged C` is the heap instance and the stack record is some magical thing users should not have to know about. From what you were trying to do it seems like that one would be a lot more usable. The problem I have with this is I don't think it would extend well at all to an `owned` or `shared` `C`, because I think it would run afoul of the memory management we're doing with the stack record instrumentation. But maybe that isn't a problem at all; we could just assume that anyone taking a class instance from Chapel->C->Chapel would know not to use Chapel memory management for that, or even give an error for `c_ptrTo` on a class type.

**Comment by mppf** [comment]
I'm just catching up on this issue.


> My best answer at this point might be that `c_addrOf()` potentially just returns a `c_void_ptr` for types that don't have an "obvious" C type

Brad, in the OP did you mean `c_ptrTo` here? Either way that is the direction [Anna proposed](https://github.com/chapel-lang/chapel/issues/22478#issuecomment-1579163561) (and had some support).

> To me, this feels consistent with the direction we went with strings and bytes. However, I'm not quite sure what the use case is for a `c_ptr(unmanaged C)`?

I think it's appropriate for `c_addrOf(myUnmanagedC)` to produce such a thing. It's really important that `c_addrOf` can take the address of anything and do so in a manner similar to what C code would do. There are certainly some C APIs that expect a pointer-to-a-pointer for one reason or other.

Now back to the big picture:

I was feeling really confused about what was happening here so I made a code example. I'm pretty sure this is not relevant for most of you who have been discussing this so far, but it might help others who come to the issue.

``` chapel
use CTypes;

class C {}
var myC: unmanaged C = new unmanaged C();

writeln("myC:c_void_ptr = ", myC: c_void_ptr);
writeln("c_ptrTo(myC) = ", c_ptrTo(myC): c_void_ptr,
        " : ", (c_ptrTo(myC)).type:string);
writeln("c_addrOf(myC) = ", c_addrOf(myC): c_void_ptr,
        " : ", (c_addrOf(myC).type):string);
```

On `main` today, when you compile this without flags, you get some warnings about `c_ptrTo(class)` changing and then it has this behavior (obviously the pointer values will differ on different systems).

```
myC:c_void_ptr = 0x7f3cbc002be0
c_ptrTo(myC) = 0x5569a57a91b0 : c_ptr(unmanaged C)
c_addrOf(myC) = 0x5569a57a91b0 : c_ptr(unmanaged C)
```

OK, so `myC:c_void_ptr` gives a pointer to the instance on the heap; while `c_ptrTo(myC)` and `c_addrOf(myC)` are both pointers-to-pointers.

If we compile with the new behavior (`-s cPtrToStringBytesClassLogicalAddress=true`), here is what we get:
```
myC:c_void_ptr = 0x7f3834002be0
c_ptrTo(myC) = 0x7f3834002be0 : c_ptr(unmanaged C)
c_addrOf(myC) = 0x55d635e191b0 : c_ptr(unmanaged C)
```

Now c_ptrTo returns the instance, but casts it to `c_ptr(unmanaged C)`. `c_addrOf` is unchanged.

To me, the fact that `c_ptrTo(myC)` returns `c_ptr(unmanaged C)` with the same value as `myC:c_void_ptr` is an error. I don't think it makes sense in the type system. I have been following along (but not directly involved) with the decisions in this area but my understanding of the decision that it was to make `c_ptrTo(myC)` behave exactly the same as `myC:c_void_ptr` which would imply that the return type is `c_void_ptr` as well. ([Link to decision process details for those who can see it](https://github.com/Cray/chapel-private/issues/4824#issuecomment-1553476379))

While it is technically possible, I do not think that we should change what `c_ptr(unmanaged C)` means. I think that needs to be a pointer-to-a-pointer due to the way that we say "Values of class types are already pointers" and "The default intent for class types is `const in` but that just copies the pointer". IMO if we were to say that `c_ptr(unmanaged C)` is a type that means a pointer to the heap object, that would imply that either 1) We have the `c_ptr` type constructor doing something weird and special for classes or 2) We are moving away from "Class values are pointers" towards "Class values are references to objects that happen to be allocated on the heap" (which would imply, for example, that the default intent for a formal of class type would be `const ref`, being a reference to the object itself on the heap). I think of both (1) and (2) here as "a bridge too far" at this point.

Bottom line, what do I think we should do about it?

 1. Immediately adjust the return type of `c_ptrTo(class)` to be `c_void_ptr`. IMO this is a bug fix (because IMO the decision made was that `c_ptrTo(myC)` behaves exactly as `myC:c_void_ptr` -- therefore it should return `c_void_ptr`). However I don't know the details of that decision but at the very least this approach seems to have a good amount of agreement in the discussion here. 
 2. Consider, in the future, adjusting the return type of `c_ptrTo(class)` to be `c_ptr(ClassImplStruct)` where `ClassImplStruct` is a distinct type for each class that is created by the implementation but is not otherwise nameable. In particular, for `class C` above, at code-generation time this structure exists and is called `chpl_C_chpl_object`. Ideally we would consider this before the fall release; otherwise we would not be able to stabilize `c_ptrTo` in that release. I suspect that changing this will have modest impact on applications (because, generally speaking, you can implicitly convert `c_ptr(SomeType)` to `c_void_ptr`); however it is theoretically possible it could break something.

**Comment by riftEmber** [comment]
I agree with the view that `c_ptrTo(myC)` returning a `c_ptrTo(unmanaged C)` right now is a bug. Seeing support for changing its return type to `c_void_ptr`, I will do that today in time for the upcoming release. `c_ptrTo(myC)` is intended to work the same as `myC:c_void_ptr` did previously, and can be used to take a class from Chapel->C->Chapel as the cast was before.

This addresses the confusion coming from `c_ptrTo` and `c_addrOf` returning the same type, but not necessarily of `c_addrOf` giving you a `c_ptr(unmanaged C)` that can't be dereferenced later. I think this calls for adding information to our C Interop technote about this particular case, to explain that an `unmanaged C` is a pointer to something on the heap, so if you want a pointer to that thing on the heap you have to use the special `c_ptrTo` behavior instead of `c_addrOf`. I don't know if there's a better solution to @twesterhout 's issue that wouldn't involve some abuse of types; while at first glance a `c_ptr(unmanaged C)` being a pointer-to-pointer is confusing, it is accurate to our implementation. I don't think there's a way around getting somewhat down in the weeds of the implementation for an operation across the language barrier like this.

**Comment by riftEmber** [comment]
I have created https://github.com/chapel-lang/chapel/pull/22522 to add documentation specifically about the use of `c_ptrTo` and `c_addrOf` with class types.

**Comment by bradcray** [comment]
@riftEmber / @twesterhout / @jabraham17  : I just stumbled across this issue and am wondering whether it could be closed now, or whether there are still aspects of it that need work and attention?  If so, can you summarize what more needs to be done?

(I admittedly didn't re-read the whole issue, but am just remembering Anna had done some work here, see the mentioning commits and PRs, so am being lazy ("time-efficient"?) in asking you).  Thanks!

**Comment by riftEmber** [comment]
I think we have things in a decent state right now and all of the implementation we wanted to consider this "stable" done. However, I don't think we've addressed @mppf 's proposed later change in https://github.com/chapel-lang/chapel/issues/22478#issuecomment-1580738408:
> Consider, in the future, adjusting the return type of c_ptrTo(class) to be c_ptr(ClassImplStruct) where ClassImplStruct is a distinct type for each class that is created by the implementation but is not otherwise nameable.

**Comment by bradcray** [comment]
OK, thanks Anna!  Maybe we could fork that off to its own issue, just to simplify things (assuming @mppf is enough of a champion to warrant it?)

**Comment by mppf** [comment]
I wrote up #24161 about the idea so that this issue can be closed. That said, I'm not entirely convinced that it would be worthwhile to do. Perhaps there is some optimization benefit but I'm not seeing much other benefit unless we add a bunch of other features that don't seem critical.

**Comment by riftEmber** [comment]
Thanks Michael. Closing this until/unless someone points out something I missed
